### DISCUSSION TITLE
Support javascript config files in addition to `json`

### DISCUSSION THREAD

**Discussion by claytercek** [bluecadet/launchpad#65] [Ideas]
Would it make sense to support javascript config files (`js`/`cjs`/`mjs`) in addition to `json` for users that want more control? Or would that be too redundant with the node API?

### Env Overrides

For programs that have multiple instances that require slightly different configs, it seems the common pattern currently is to have different launchpad config files for each instance. Having a js config file would allow users to use a single file that could adjust the config based on env variables or user variables:

```js
const os = require("os");

module.exports = {
  content: {
    sources: [
      {
        type: "sanity",
        queries: [
          {
            id: "config",
            // Use some custom env variable
            query: `*[_type == 'app-config' && pcName == '${process.env.PC_NAME}']`,
          },
        ],
      },
    ],
  },
  monitor: {
    apps: [
      {
        pm2: {
          name: "client",
          // Or access some node built-in variables
          cwd: `${os.homedir()}\\AppData\\Local\\Google\\Chrome\\Application`,
          script: "chrome.exe",
        },
        // ...
      },
    ],
  },
  // ...
};
```

### Sharing Config Code

Having js files would allow us to import other files into our config files:

```js

const someSharedContentConfig = require(".../content.json");

module.exports = {
   // ...
   content: someSharedContentConfig
}

// we could also export some utility functions from launchpad that could make common setups easier:

const chromeApp = require("@bluecadet/launchpad-monitor/chromeApp");

module.exports = {
  // ...
  monitor: chromeApp() // or chromeApp({...overrides})
}
```

### Jsdoc / typescript Hints

Using js for config also means we could take advantage of the JSDoc hints while writing up config files, similar to how nextjs types their config files: 

```js
/**
 * @type {import('next').NextConfig}
 */
const nextConfig = {
  /* config options here */
}

module.exports = nextConfig
```

--- 

If this all is too redundant with the node API, then it might make more sense to just show how to accomplish these same scenarios using the node API somewhere in the documentation. 

There are a couple of existing tools that will automagically search for config files with different file types, ex: https://github.com/davidtheclark/cosmiconfig

**Comment by benjaminbojko** [comment]
Yes!!! This actually solves a lot of issues and it would actually not require much work at all.

As a matter of fact, it's already supported! The `ConfigManager` class already has a `importJsConfig()` function:

https://github.com/bluecadet/launchpad/blob/f43492d7d106adbd4c4bad829ad56d8a2121faa1/packages/utils/lib/config-manager.js#L51-L75

And the `npm run test` script already uses it

https://github.com/bluecadet/launchpad/blob/f43492d7d106adbd4c4bad829ad56d8a2121faa1/packages/launchpad/test/test.js#L4-L6

https://github.com/bluecadet/launchpad/blob/f43492d7d106adbd4c4bad829ad56d8a2121faa1/packages/launchpad/test/config.js#L1-L15

In fact, we need this for a project now and I might just shimmy in a quick solution/test and then we can see if we can incorporate something like cosmiconfig, which sounds super helpful and would also address #7 

The only challenge might be that we're currently using [yargs](https://github.com/yargs/yargs-parser#configuration) for loading, which also supports overriding via CLI params. That seems solvable though.

**Reply by claytercek** [reply]
Oh awesome! So if I'm understanding this correctly, the ConfigManager has it as a static utility fn for use with the node API, and all we really need to do is find the right spot to incorporate it when launching launchpad from the CLI?

**Reply by benjaminbojko** [reply]
That's exactly right! I may actually need to look into that today and will keep you posted.

**Comment by claytercek** [comment]
Resolved with #124, plus intellisense is being added with #128
